mmrotate/apis/inference.py

```diff
@@ -1,94 +1,112 @@
-        steps (list): The steps between two patches.
-        ratios (list): Image resizing ratios for multi-scale detecting.
-        merge_iou_thr (float): IoU threshold for merging results.
+        imgs (str, ndarray, Sequence[str/ndarray]): Either image files or
+            loaded images.
+        sizes (list[int]): The sizes of patches.
+        steps (list[int]): The steps between two patches.
+        ratios (list[float]): Image resizing ratios for multi-scale detecting.
+        nms_cfg (dict): nms config.
         bs (int): Batch size, must greater than or equal to 1.
 
     Returns:
         list[np.ndarray]: Detection results.
     """
     assert bs >= 1, 'The batch size must greater than or equal to 1'
+    if isinstance(imgs, (list, tuple)):
+        is_batch = True
+    else:
+        imgs = [imgs]
+        is_batch = False
     cfg = model.cfg
-    device = next(model.parameters()).device  # model device
-    cfg = cfg.copy()
-    # set loading pipeline type
-    cfg.data.test.pipeline[0].type = 'LoadPatchFromImage'
-    cfg.data.test.pipeline = replace_ImageToTensor(cfg.data.test.pipeline)
-    test_pipeline = Compose(cfg.data.test.pipeline)
-
-    if not isinstance(img, np.ndarray):
-        img = mmcv.imread(img)
-    height, width = img.shape[:2]
-    sizes, steps = get_multiscale_patch(sizes, steps, ratios)
-    windows = slide_window(width, height, sizes, steps)
-
-    results = []
-    start = 0
-    while True:
-        # prepare patch data
-        patch_datas = []
-        if (start + bs) > len(windows):
-            end = len(windows)
-        else:
-            end = start + bs
-        for window in windows[start:end]:
-            data = dict(img=img, win=window.tolist())
-            data = test_pipeline(data)
-            patch_datas.append(data)
-        data = collate(patch_datas, samples_per_gpu=len(patch_datas))
-        # just get the actual data from DataContainer
-        data['img_metas'] = [
-            img_metas.data[0] for img_metas in data['img_metas']
-        ]
-        data['img'] = [img.data[0] for img in data['img']]
-        if next(model.parameters()).is_cuda:
-            # scatter to specified GPU
-            data = scatter(data, [device])[0]
-        else:
-            for m in model.modules():
-                assert not isinstance(
-                    m, RoIPool
-                ), 'CPU inference with RoIPool is not supported currently.'
-
-        # forward the model
-        with torch.no_grad():
-            results.extend(model(return_loss=False, rescale=True, **data))
-
-        if end >= len(windows):
-            break
-        start += bs
-
-    results = merge_results(
-        results,
-        windows[:, :2],
-        img_shape=(width, height),
-        iou_thr=merge_iou_thr,
-        device=device)
-    return results
+
+    if test_pipeline is None:
+        cfg = cfg.copy()
+        test_pipeline = cfg.test_dataloader.dataset.pipeline
```

Comment:
What if there is a dataset wrapper?

Reply:
I add a new function named `get_test_pipeline_cfg` to handle the special cases of dataset wrappers. mmdetection also has the same problem.